marc.py

```diff
@@ -592,52 +592,91 @@ def create_record(cls, work, annotator, force_create=False, integration=None):
+          works for a single query.
+        """
+
+        # We mirror the content, if it's not empty. If it's empty, we create a CachedMARCFile
+        # and Representation, but don't actually mirror it.
+        if not mirror:
+            storage_protocol = self.integration.setting(self.STORAGE_PROTOCOL).value
+            if storage_protocol == ExternalIntegration.S3:
+                mirror_integration = ExternalIntegration.lookup(
+                    self._db, storage_protocol, ExternalIntegration.STORAGE_GOAL)
+                if mirror_integration:
+                    mirror = mirror or S3Uploader(mirror_integration)
+
+        if not mirror:
+            raise Exception("No mirror integration for configured protocol %s" % storage_protocol)
+
+        # End time is before we start the query, because if any records are changed
+        # during the processing we may not catch them, and they should be handled
+        # again on the next run.
+        end_time = datetime.datetime.utcnow()
 
         works_q = lane.works(self._db)
         if start_time:
             works_q = works_q.filter(MaterializedWorkWithGenre.last_update_time>=start_time)
-        for work in works_q:
-            record = self.create_record(
-                work, annotator, force_refresh, self.integration)
-            if record:
-                output.write(record.as_marc())
 
-        content = output.getvalue()
-        output.close()
-
-        end_time = datetime.datetime.utcnow()
-
-        # Mirror the content, if it's not empty. If it's empty, create a CachedMARCFile
-        # and Representation, but don't actually mirror it.
-        storage_protocol = self.integration.setting(self.STORAGE_PROTOCOL).value
-        if storage_protocol == ExternalIntegration.S3:
-            mirror_integration = ExternalIntegration.lookup(
-                self._db, storage_protocol, ExternalIntegration.STORAGE_GOAL)
-            if not mirror_integration:
-                raise Exception("No mirror integration for configured protocol %s" % storage_protocol)
-
-            mirror = mirror or S3Uploader(mirror_integration)
-            url = mirror.marc_file_url(self.library, lane, end_time, start_time)
-            representation, ignore = get_one_or_create(
-                self._db, Representation, url=url, media_type=Representation.MARC_MEDIA_TYPE)
-            representation.fetched_at = end_time
-            representation.content = content
+        total = works_q.count()
+        offset = 0
+
+        url = mirror.marc_file_url(self.library, lane, end_time, start_time)
+        representation, ignore = get_one_or_create(
+            self._db, Representation, url=url, media_type=Representation.MARC_MEDIA_TYPE)
+
+        with mirror.multipart_upload(representation, url) as upload:
+            output = StringIO()
+            current_count = 0
+            while offset < total:
```

Comment:
I've spent some time mentally fiddling around with the idea of turning this into a generator and passing the generator into `mirror`. But what I'm actually thinking is that I'd rather see the code inside this loop in a helper method that can be tested separately.

Your existing test coverage is fine, so don't worry aobut this.